From acc31a18b26b5055b10295a05c962c8c31cb8eaf Mon Sep 17 00:00:00 2001
From: Oliver Neukum <oneukum@suse.de>
Date: Wed, 14 Aug 2013 08:51:46 +0200
Subject: [PATCH] xhci: directly calling _PS3 on suspend
Patch-Mainline: Never (Generic solution in upstream)
References: bnc#833148
Git-Commit: acc31a18b26b5055b10295a05c962c8c31cb8eaf

From external information:
I think the Linux auto wake up issue is caused by XHCI_BAR + 0x8154[31] setting.
On win7 OS, Intel XCHI driver will call _PS3 method under XHCI namespace to clear XHCI_BAR + 0x8154[31] before going to sleep state.
However, on Linux OS, Linux XHCI driver does not call OSC/PS0/PS3 method. This is problem.

Upstream this is solved by a generic PM system. Our kernel is too old.
We need to call directly.

Signed-off-by: Oliver Neukum <oneukum@suse.de>
---
 drivers/usb/host/xhci-pci.c | 16 ++++++++++++++++
 1 file changed, 16 insertions(+)

diff --git a/drivers/usb/host/xhci-pci.c b/drivers/usb/host/xhci-pci.c
index eed395a..0c88ea3 100644
--- a/drivers/usb/host/xhci-pci.c
+++ b/drivers/usb/host/xhci-pci.c
@@ -22,6 +22,8 @@
 
 #include <linux/pci.h>
 #include <linux/slab.h>
+#include <linux/acpi.h>
+#include <acpi/actypes.h>
 
 #include "xhci.h"
 
@@ -219,6 +221,13 @@ static void xhci_pci_remove(struct pci_dev *dev)
 static int xhci_pci_suspend(struct usb_hcd *hcd, bool do_wakeup)
 {
 	struct xhci_hcd	*xhci = hcd_to_xhci(hcd);
+	struct pci_dev          *pdev = to_pci_dev(hcd->self.controller);
+
+	#ifdef CONFIG_ACPI
+	acpi_handle handle = DEVICE_ACPI_HANDLE(&pdev->dev);
+	acpi_status status;
+	#endif
+
 	int	retval = 0;
 
 	if (hcd->state != HC_STATE_SUSPENDED ||
@@ -227,6 +236,13 @@ static int xhci_pci_suspend(struct usb_hcd *hcd, bool do_wakeup)
 
 	retval = xhci_suspend(xhci);
 
+	#ifdef CONFIG_ACPI
+	dev_printk(KERN_INFO, &pdev->dev, "%s try to enter PS3\n", __FUNCTION__);
+	status = acpi_bus_set_power(handle, ACPI_STATE_D3);
+	if (ACPI_FAILURE(status))
+	    dev_printk(KERN_ERR, &pdev->dev, "%s FAILED to enter PS3\n", __FUNCTION__);
+	#endif
+
 	return retval;
 }
 
-- 
1.8.3.1

